src/theory/quantifiers/sygus/sygus_reconstruct.cpp

```diff
@@ -0,0 +1,573 @@
+  {
+    builtinVars.push_back(datatypes::utils::sygusToBuiltin(sv));
+    d_groundVars.emplace(datatypes::utils::sygusToBuiltin(sv),
+                         datatypes::utils::sygusToBuiltin(sv));
+  }
+
+  SygusTypeInfo stnInfo;
+  stnInfo.initialize(d_tds, stn);
+
+  // find the non-terminals of the grammar
+  std::vector<TypeNode> sfTypes;
+  stnInfo.getSubfieldTypes(sfTypes);
+
+  // initialize the enumerators and candidate rewrite databases. Notice here
+  // that we treat the sygus variables introduced by the problem as bound
+  // variables (needed for making sure that obligations are equal). This is fine
+  // as we will never add variables that were introduced by the enumerators to
+  // the database.
+  for (TypeNode tn : sfTypes)
+  {
+    d_stnInfo[tn].initialize(d_tds, d_stats, tn, builtinVars);
+  }
+}
+
+void SygusReconstruct::removeSolvedObs(
+    std::unordered_map<TypeNode,
+                       std::unordered_set<Node, NodeHashFunction>,
+                       TypeNodeHashFunction>& unsolvedObs)
+{
+  for (std::pair<const TypeNode, std::unordered_set<Node, NodeHashFunction>>&
+           tempPair : unsolvedObs)
+  {
+    std::unordered_set<Node, NodeHashFunction>::iterator it =
+        tempPair.second.begin();
+    while (it != tempPair.second.end())
+    {
+      if (d_sol[*it] == Node::null())
+      {
+        ++it;
+      }
+      else
+      {
+        it = tempPair.second.erase(it);
+      }
+    }
+  }
+}
+
+Node SygusReconstruct::mkGround(Node n) const
+{
+  // get the set of bound variables in n
+  std::unordered_set<TNode, TNodeHashFunction> vars;
+  expr::getVariables(n, vars);
+
+  std::unordered_map<TNode, TNode, TNodeHashFunction> subs;
+
+  // generate a ground value for each one of those variables
+  for (const TNode& var : vars)
+  {
+    subs.emplace(var, var.getType().mkGroundValue());
+  }
+
+  // substitute the variables with ground values
+  return n.substitute(subs);
+}
+
+bool SygusReconstruct::notify(Node s,
+                              Node n,
+                              std::vector<Node>& vars,
+                              std::vector<Node>& subs)
+{
+  for (size_t i = 0; i < vars.size(); ++i)
+  {
+    // We consider sygus variables as ground terms. So, if they are not equal to
+    // their substitution, then s is not matchable with n and we try the next
+    // term s Example: If s = (+ z x) and n = (+ z y), then s is not matchable
+    // with n and we return true
+    if (d_groundVars.find(vars[i]) != d_groundVars.cend() && vars[i] != subs[i])
+    {
+      return true;
+    }
+  }
+  // Note: false here means that we finally found an s that is matchable with n,
+  // so we should not add n to the pool
+  return false;
+}
+
+void SygusReconstruct::clear()
+{
+  d_obInfo.clear();
+  d_stnInfo.clear();
+  d_sol.clear();
+  d_subObs.clear();
+  d_parentOb.clear();
+  d_groundVars.clear();
+  d_poolTrie.clear();
+}
+
+std::string SygusReconstruct::ob(Node k) const
```

Comment:
Can you move this method to the obligation utility class?

Reply:
Done. Looks a bit uglier though.